third_party/nvfuser/csrc/multidevice/aggregate_dag.cpp

```diff
@@ -0,0 +1,211 @@
+
+    for (auto output_val : group->output_vals) {
+      // creates the AggregateVal corresponding to each output of the current
+      // group
+      auto val = IrBuilder::create<AggregateVal>(container, output_val, group);
+      // tag current group as producer of this val
+      producer[output_val] = val;
+      // add the new IR to the outputs container
+      outputs.push_back(val);
+    }
+
+    for (auto input_val : group->input_vals) {
+      // creates the AggregateVal corresponding to each input of the current
+      // group
+      auto val = IrBuilder::create<AggregateVal>(container, input_val, group);
+      // add current group as a consumer of this val
+      consumers.insert({input_val, val});
+      // add the new IR to the inputs container
+      inputs.push_back(val);
+      if (producer.find(input_val) != producer.end()) {
+        // means that input_val is not a global input and
+        // so is produced by another group
+        auto src = producer[input_val];
+        // create and IR indicating the val needs to be sent between groups
+        auto sendRecv = IrBuilder::create<SendRecv>(container, val, src);
+        // the received val is set as output of sendRecv
+        val->setDefinition(sendRecv);
+      } else {
+        // add val as global input of the aggregate dag
+        addInput(val);
+      }
+    }
+
+    // create IR representing the exprs contained in the current group
+    // This could be moved to AggregateExpr's initializer.
+    // Question: what does the following mean? "Constructors need to register
+    // with the Fusion after inputs/outputs are defined"
+    auto expr = IrBuilder::create<AggregateExpr>(container, group);
+    for (auto out : outputs) {
+      // set each output as an output of the group
+      expr->addOutput(out);
+      // set definition of the ouput
+      out->setDefinition(expr);
+    }
+    for (auto in : inputs) {
+      // set each input as an input of the group
+      expr->addInput(in);
+    }
+  }
+  for (auto it : producer) {
+    if (consumers.count(it.first) == 0) {
+      // means that it is a global output.
+      // TODO: Im not sure it is exact, but it is not critical for now
+      addOutput(it.second);
+    }
+  }
+}
+
+  void AggregateDag::print() {
+    AggregateDagPrinter p(this);
+    p.print();
+  }
+
+
+AggregateExpr::AggregateExpr(IrBuilderPasskey passkey, GroupPtr group)
+    : Expr(passkey), group_(group) {
+  TORCH_INTERNAL_ASSERT(
+      passkey.ir_container_->isA<AggregateDag>(),
+      "IR type only valid for AggregateDag container.");
+}
+
+std::string AggregateExpr::toString(int indent_size) const {
+  std::stringstream ss;
+  ss  << "AggregateExpr representing Group " << getGroup()->unique_id<<".";
+  ss << "Inputs={";
+  for (auto input: inputs()){
+    //TODO: use dispatcher to print the Original Val (here, we assume it is a TensorView)
+    ss << input->as<AggregateVal>()->getOriginalVal()->as<TensorView>()->toString(indent_size);
```

Comment:
Is there a nicer way to print the Val than explicitly casting it to TensorView?
I would like this function to print even if the Val is not a TensorView
So need to use some kind of dispatch -- I don't understand how it is done in the new interface

Reply:
Not sure what you're asking, but `toString` is a virtual function, so it shouldn't be necessary to cast it to `TensorView`.

Reply:
changed